src/thumbnailer.cc

```diff
@@ -150,4 +146,119 @@ Thumbnailer::Status Thumbnailer::GenerateAnimation(WebPData* const webp_data) {
   return kOk;
 }
 
+Thumbnailer::Status Thumbnailer::GenerateAnimationEqualPSNR(
+    WebPData* const webp_data) {
+  // Rearrange frames.
+  std::sort(frames_.begin(), frames_.end(),
+            [](const FrameData& A, const FrameData& B) -> bool {
+              return A.timestamp_ms < B.timestamp_ms;
+            });
+
+  int final_psnr = -1;
+  WebPData new_webp_data;
+  WebPDataInit(&new_webp_data);
+  WebPPicture pic;
+  WebPPictureInit(&pic);
+  WebPAuxStats stats;
+
+  for (int target_psnr = 99; target_psnr > 0; --target_psnr) {
+    std::cerr << "Target PSNR: " << target_psnr << ". ";
+    bool ok = true;
+
+    WebPAnimEncoderDelete(enc_);
+    enc_ = WebPAnimEncoderNew((frames_[0].pic).width, (frames_[0].pic).height,
+                              &anim_config_);
+    if (enc_ == nullptr) ok = 0;
+
+    // For each frame, find the quality value that produces WebPPicture
+    // having PSNR close to target_psnr.
+    if (ok)
+      for (auto& frame : frames_) {
+        int min_quality = 0;
+        int max_quality = 100;
+        int final_quality = -1;
+
+        // Get lowest PSNR value possible.
+        WebPPictureInit(&pic);
+        WebPPictureCopy(&frame.pic, &pic);
+        pic.stats = &stats;
+        config_.quality = min_quality;
+        WebPEncode(&config_, &pic);
+        int low_psnr = pic.stats->PSNR[3];
+        WebPPictureFree(&pic);
+
+        // Get highest PSNR value possible.
+        WebPPictureInit(&pic);
+        WebPPictureCopy(&frame.pic, &pic);
+        pic.stats = &stats;
+        config_.quality = max_quality;
+        WebPEncode(&config_, &pic);
+        int high_psnr = pic.stats->PSNR[3];  // PSNR-all.
+        WebPPictureFree(&pic);
+
+        // Target PSNR is out of range.
+        if (target_psnr > high_psnr || target_psnr < low_psnr) {
+          ok = false;
+          break;
+        }
```

Comment:
If the target_psnr is outside of the range, I think we should fall back to using the closest value anyway (so use min_quality or max_quality) instead of returning an error

Reply:
It looks like you haven't addressed this comment. If you disagree (which is fine!) or if you want to fix this in another pull request, please explain that in a comment.

Reply:
If I used min_quality or max_quality instead, `thumbnailer` might successfully generate a animation that is not satisfactory (psnr values not approximately close to each other). Anyway, this is now a rare case since I reduced the search size.